### DISCUSSION TITLE
Understanding Laravel JSON encoding of floats

### DISCUSSION THREAD

**Discussion by 3cooper** [laravel/framework#48311] [Q&A]
I have the float `-30.22058000` stored in the database and accessed via a model.  When I return an array of those models in JSON, the float is turned into 
`-30.220579999999998`.  In a `JsonResource`, I hardcoded the value to the original `-30.22058000` instead of returning the property of the model.  The response now contains `-30.22058`.   If I dump the property via `dump($this->myFloat)` or `var_dump($this->myFloat)`, I see the float `-30.22058`.  I'm trying to figure out what causes the difference in the JSON encoded response between when I hardcode the float value vs pull it from the database.  Is it possible that the value in the property is not really what is being dumped, and dump is showing an approximation?  There are no casts configured for this model.  My `serialize_precision` is set to `-1` in my `php.ini` file.  I've tried to trace this through the JSON serialization and I can also see the float value is still as expected at the end `JsonResource::resolve`.  This would lead me to believe the same value is going into the actual `json_encode` function.  Any insight on this would be great.

My solution for now, will be in the resource to return `floatval(number_format($this->myFloat, 8))`.  This seems weird since I'm still returning a float, but the value for this is the shorter form of `-30.22058`.

**Comment by Angel5a** [comment]
Your database float is only 4 bytes in length and your PHP float is 8 bytes in length. Create a double column, you'll get a "correct" value. But both types of real numbers are just an approximations.

More details on wiki [IEEE_754](https://en.wikipedia.org/wiki/IEEE_754).

**Comment by 3cooper** [comment]
I checked the database, and the column is defined as `double(10,8)` which I believe would mean it is 8 bytes.  Am I correct with that?  Even with that, when dumping the data in PHP, I was presented with the same value in the dump even though encoding that value in JSON led to different values.

**Reply by Angel5a** [reply]
> MySQL permits a nonstandard syntax: FLOAT(M,D) or REAL(M,D) or DOUBLE PRECISION(M,D). Here, (M,D) means than values can be stored with up to M digits in total, of which D digits may be after the decimal point. 

So, when you put your number into database, it gets rounded to 10 significant digits (float < about 32 significant bits < double). When you read in back, you get approximate floating point value of this 10 digits. Floating point values are approximate values (no only in MySQL, but by nature). You should not compare them with strict equation, but with epsilon. And all this topic is about a problem can be encountered.

What you see is OK in terms of using floating point numbers. If you need a nice shiny values in you json responses, then you need some workarounds, like one, you posted, or any other rounding to significant digits.

**Comment by 3cooper** [comment]
I used `rtrim(number_format($this->myFloat, 100),0)` to dump the value instead of just `dump` or `var_dump`.  This gave me a value of `-30.2205799999999982219378580339252948760986328125`.  When I dump the literal value using the same method, I get `-30.220580000000001774651536834426224231719970703125`.   I have limited experience with floats.  My understanding from this is that starting at database, I am not seeing the actual value being used and even when dumping it, it is still being rounded in most cases.
